src/mpi4py/util/pkl5.py

```diff
@@ -0,0 +1,534 @@
+        """Create and return a new object."""
+        if message is None:
+            message = (MPI.MESSAGE_NULL,)
+        if isinstance(message, MPI.Message):
+            message = (message,)
+        return super().__new__(cls, message)
+
+    def __eq__(self, other):
+        """Return ``self==other``."""
+        if isinstance(other, Message):
+            return tuple(self) == tuple(other)
+        if isinstance(other, MPI.Message):
+            return all(msg == other for msg in self)
+        return NotImplemented
+
+    def __ne__(self, other):
+        """Return ``self!=other``."""
+        if isinstance(other, Message):
+            return tuple(self) != tuple(other)
+        if isinstance(other, MPI.Message):
+            return any(msg != other for msg in self)
+        return NotImplemented
+
+    def __bool__(self):
+        """Return ``bool(self)``."""
+        return any(msg for msg in self)
+
+    def recv(self, status=None):
+        """Blocking receive of matched message."""
+        return _mrecv(self, status)
+
+    def irecv(self):
+        """Nonblocking receive of matched message."""
+        return _imrecv(self)
+
+    @classmethod
+    def probe(cls, comm,
+              source=ANY_SOURCE, tag=ANY_TAG,
+              status=None):
+        """Blocking test for a matched message."""
+        return _mprobe(comm, MPI.Comm.Mprobe, source, tag, status)
+
+    @classmethod
+    def iprobe(cls, comm,
+               source=ANY_SOURCE, tag=ANY_TAG,
+               status=None):
+        """Nonblocking test for a matched message."""
+        return _mprobe(comm, MPI.Comm.Improbe, source, tag, status)
+
+
+class Comm(MPI.Comm):
+    """Communicator."""
+
+    def send(self, obj, dest, tag=0):
+        """Blocking send in standard mode."""
+        _send(self, MPI.Comm.Send, obj, dest, tag)
+
+    def bsend(self, obj, dest, tag=0):
+        """Blocking send in buffered mode."""
+        _send(self, MPI.Comm.Bsend, obj, dest, tag)
+
+    def ssend(self, obj, dest, tag=0):
+        """Blocking send in synchronous mode."""
+        _send(self, MPI.Comm.Ssend, obj, dest, tag)
+
+    def isend(self, obj, dest, tag=0):
+        """Nonblocking send in standard mode."""
+        return _isend(self, MPI.Comm.Isend, obj, dest, tag)
+
+    def ibsend(self, obj, dest, tag=0):
+        """Nonblocking send in buffered mode."""
+        return _isend(self, MPI.Comm.Ibsend, obj, dest, tag)
+
+    def issend(self, obj, dest, tag=0):
+        """Nonblocking send in synchronous mode."""
+        return _isend(self, MPI.Comm.Issend, obj, dest, tag)
+
+    def recv(self,
+             buf=None, source=ANY_SOURCE, tag=ANY_TAG,
+             status=None):
+        """Blocking receive."""
+        return _recv(self, MPI.Comm.Recv, buf, source, tag, status)
+
+    def irecv(self,
+              buf=None, source=ANY_SOURCE, tag=ANY_TAG):
+        """Nonblocking receive."""
+        raise RuntimeError("unsupported")
```

Comment:
I could not figure out a way. The problem is that to issue an `MPI_Irecv()` call you need an upper bound of the message size. 

The normal `irecv()` method uses (by default) a 32K buffer (our you pass a buffer size, or your own scratch buffer space) as a guess, and if your guess is wrong, MPI will complain about a truncated message. Perhaps I should try do something similar here. I think when I started I figured out there was an issue with that approach in here, but maybe I was wrong. Let me think a bit more about it.

Reply:
The most silly workaround I can think of is to do a blocking send/recv to send the buffer size, in the hope that the overhead is amortized by subsequent large, nonblocking transfer, but I am sure you will think of a better way 😅

But, again, the feature is complete and we can always revisit later. This question should be nonblocking too. We just need to [document](https://github.com/mpi4py/mpi4py/pull/29#discussion_r593377553) this.

Reply:
No, no, these tricks do not work reliably! A nonblocking operation has to be truly nonblocking, otherwise you may deadlock. For example, look at the following two pieces of code:

```python
# version 1
comm.ssend(sobj, dest=comm.rank)
req = comm.irecv(source=comm.rank)
robj = req.wait()
```

```python
# version 2
req = comm.irecv(source=comm.rank)
comm.ssend(sobj, dest=comm.rank)
robj = req.wait()
```

A proper implementation of `irecv()` should be such that both versions work as expected. But if you add an initial blocking step, then both will deadlock. In short, better safe than sorry, do not promise what you cannot deliver.

PS: Now I remembered all the issues. `irecv()` cannot possibly be implemented without using polling on a separate thread, and even if I were willing do it, then I cannot make things thread-safe, other users threads would compete to receive messages, completely screwing things up, and there is no way to use a lock without bloking other `recv()` operations (and then,  what the point of using nonblocking receive would be?).

Reply:
You are absolutely right. Having blocking send/recv would break the semantics and cause deadlocks. For some reason I kept thinking this blocking pair can be hidden safely somewhere...